[custom_op] support default dtype values

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129189
* #129179
* #129178
* #129177

This PR:
- moves some of the dtype-string utilities into ScalarType.{h, cpp}
- adds a new utility to get a mapping from dtype name to the C++ dtype
- the perser now checks if the string is a dtype name; if it is then it
  pulls the c++ dtype from the mapping.

Test Plan:
- new tests

zou3519 committed:
[custom_op] support default dtype values

This PR:
- moves some of the dtype-string utilities into ScalarType.{h, cpp}
- adds a new utility to get a mapping from dtype name to the C++ dtype
- the perser now checks if the string is a dtype name; if it is then it
  pulls the c++ dtype from the mapping.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "[custom_op] support default dtype values"


This PR:
- moves some of the dtype-string utilities into ScalarType.{h, cpp}
- adds a new utility to get a mapping from dtype name to the C++ dtype
- the perser now checks if the string is a dtype name; if it is then it
  pulls the c++ dtype from the mapping.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "[custom_op] support default dtype values"


This PR:
- moves some of the dtype-string utilities into ScalarType.{h, cpp}
- adds a new utility to get a mapping from dtype name to the C++ dtype
- the perser now checks if the string is a dtype name; if it is then it
  pulls the c++ dtype from the mapping.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "[custom_op] support default dtype values"


This PR:
- moves some of the dtype-string utilities into ScalarType.{h, cpp}
- adds a new utility to get a mapping from dtype name to the C++ dtype
- the perser now checks if the string is a dtype name; if it is then it
  pulls the c++ dtype from the mapping.

Test Plan:
- new tests

[ghstack-poisoned]

albanD commented on c10/core/ScalarType.cpp:
```diff
@@ -128,4 +128,110 @@ ScalarType promoteTypes(ScalarType a, ScalarType b) {
+    return result;
+  }
+
+#define DEFINE_SCALAR_TYPE(_1, n) c10::ScalarType::n,
```
undef after use if it is not expected to be used elsewhere

zou3519 replied:
sg, I moved this code over from somewhere else

albanD replied:
Ho it is actually copied, not moved. Should the macro be only defined once and re-used both places then?

albanD commented on torch/csrc/jit/frontend/function_schema_parser.cpp:
```diff
@@ -219,18 +220,14 @@ struct SchemaParser {
           return static_cast<int64_t>(at::Reduction::Mean);
         } else if ("contiguous_format" == text) {
           return static_cast<int64_t>(c10::MemoryFormat::Contiguous);
+        } else if (str2dtype.count(text) > 0) {
```
Could you clarify the implication of this change? I'm not familiar with when this code is being used.

zou3519 replied:
This function gets called during construction of a FunctionSchema (https://www.internalfb.com/code/fbsource/[aca8f430d0152e7ffe6b24aedb8ee29808a16e3f]/fbcode/caffe2/torch/csrc/jit/frontend/function_schema_parser.cpp?lines=200%2C202%2C285%2C314%2C327): it is responsible for parsing the default values in the schema and turning them into an IValue. The implication of the change is now we support specifying the default value to be any of the keys in the str2dtype mapping (like "float64", "complex128", etc).

albanD replied:
Ok!
I guess none of these will be ambiguous!

zou3519 replied:
There's a bug here though, we only want to parse this as a dtype if we're sure the type is a dtype, otherwise weird things like "int x = complex128" will work... so I'll try to fix that. The bug is pre-existing though

albanD replied:
Yeah I guess anything that gets silently converted to int on the way down can have anything as it's default haha

zou3519 committed:
Update on "[custom_op] support default dtype values"


This PR:
- moves some of the dtype-string utilities into ScalarType.{h, cpp}
- adds a new utility to get a mapping from dtype name to the C++ dtype
- the perser now checks if the string is a dtype name; if it is then it
  pulls the c++ dtype from the mapping.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "[custom_op] support default dtype values"


This PR:
- moves some of the dtype-string utilities into ScalarType.{h, cpp}
- adds a new utility to get a mapping from dtype name to the C++ dtype
- the perser now checks if the string is a dtype name; if it is then it
  pulls the c++ dtype from the mapping.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge
